protocols/autonat/src/behaviour.rs

```diff
@@ -0,0 +1,638 @@
+    }
+
+    fn inject_dial_failure(
+        &mut self,
+        peer_id: Option<PeerId>,
+        handler: Self::ProtocolsHandler,
+        error: &DialError,
+    ) {
+        self.inner.inject_dial_failure(peer_id, handler, error);
+        if let Some((_, channel)) = peer_id.and_then(|p| self.ongoing_inbound.remove(&p)) {
+            // Failed to dial any of the addresses sent by the remote peer in their dial-request.
+            let _ = self
+                .inner
+                .send_response(channel, DialResponse::Err(ResponseError::DialError));
+        }
+    }
+
+    fn inject_listen_failure(
+        &mut self,
+        local_addr: &Multiaddr,
+        send_back_addr: &Multiaddr,
+        handler: Self::ProtocolsHandler,
+    ) {
+        self.inner
+            .inject_listen_failure(local_addr, send_back_addr, handler)
+    }
+
+    fn inject_new_listener(&mut self, id: ListenerId) {
+        self.inner.inject_new_listener(id)
+    }
+
+    fn inject_new_listen_addr(&mut self, id: ListenerId, addr: &Multiaddr) {
+        self.inner.inject_new_listen_addr(id, addr);
+    }
+
+    fn inject_expired_listen_addr(&mut self, id: ListenerId, addr: &Multiaddr) {
+        self.inner.inject_expired_listen_addr(id, addr);
+    }
+
+    fn inject_listener_error(&mut self, id: ListenerId, err: &(dyn std::error::Error + 'static)) {
+        self.inner.inject_listener_error(id, err)
+    }
+
+    fn inject_listener_closed(&mut self, id: ListenerId, reason: Result<(), &std::io::Error>) {
+        self.inner.inject_listener_closed(id, reason)
+    }
+
+    fn inject_new_external_addr(&mut self, addr: &Multiaddr) {
+        self.inner.inject_new_external_addr(addr);
+    }
+
+    fn inject_expired_external_addr(&mut self, addr: &Multiaddr) {
+        self.inner.inject_expired_external_addr(addr);
+    }
+
+    fn poll(
+        &mut self,
+        cx: &mut Context<'_>,
+        params: &mut impl PollParameters,
+    ) -> Poll<NetworkBehaviourAction<Self::OutEvent, Self::ProtocolsHandler>> {
+        loop {
+            match self.inner.poll(cx, params) {
+                Poll::Ready(NetworkBehaviourAction::GenerateEvent(event)) => match event {
+                    RequestResponseEvent::Message { peer, message } => match message {
+                        RequestResponseMessage::Request {
+                            request_id: _,
+                            request,
+                            channel,
+                        } => match self.handle_request(peer, request) {
+                            Some(addrs) => {
+                                self.ongoing_inbound.insert(peer, (addrs, channel));
+                                return Poll::Ready(NetworkBehaviourAction::DialPeer {
+                                    peer_id: peer,
+                                    handler: self.inner.new_handler(),
+                                    condition: DialPeerCondition::Always,
+                                });
```

Comment:
https://github.com/libp2p/rust-libp2p/pull/2317 just merged. I guess the tracking of `addrs` in `ongoing_inbound` is no longer needed?

Reply:
With https://github.com/libp2p/rust-libp2p/pull/2262/commits/310b086423e243696b23b0afc3751d2242999b8c it now specifies a fixed set of addresses. But the addresses still need to be tracked in `ongoing_inbound` to verify on `inject_connection_established` that the address we connected to was actually one that we were supposed to dial.